### PR TITLE
#862 - Remove node enhancements

### PR DESCRIPTION
**Pull request by kvanhijf**

### CODE REVIEW

File: ovs/lib/setup.py [hunk 1]
```diff
@@ -581,165 +580,202 @@ def promote_or_demote_node(node_action, cluster_ip=None):
...
-                slave_ips = [sr.ip for sr in StorageRouterList.get_slaves()]
-                offline_node_ips = [node.ip for node in storage_routers_offline]
-                SetupController._restart_framework_and_memcache_services(master_ips, slave_ips, ip_client_map, offline_node_ips)
-
-                SetupController._log(messages='    Successfully removed node\n')
+            SetupController._log(messages='Starting removal of node {0} - {1}'.format(storage_router_to_remove.name, storage_router_to_remove.ip))
+            if storage_router_to_remove_online is False:
+                SetupController._log(messages='  Marking all Storage Drivers served by Storage Router {0} as offline'.format(storage_router_to_remove.ip))
+                StorageDriverController.mark_offline(storagerouter_guid=storage_router_to_remove.guid)
+                for storagedriver in storage_router_to_remove.storagedrivers:
```
Comment (khenderick) [inline ovs/lib/setup.py]: This block is obsolete, I think

File: ovs/lib/setup.py [hunk 2]
```diff
@@ -1672,7 +1701,11 @@ def _remove_services(client, node_type):
         SetupController._log(messages='Removing services')
         services = ['workers', 'support-agent', 'watcher-framework', 'watcher-config']
         if node_type == 'master':
-            services += ['memcached', 'rabbitmq-server', 'scheduled-tasks', 'webapp-api', 'volumerouter-consumer']
+            services += ['scheduled-tasks', 'webapp-api', 'volumerouter-consumer']
+            if SetupController._is_internally_managed(service='rabbitmq'):
```
Comment (khenderick) [inline ovs/lib/setup.py]: Consistency please

File: ovs/lib/storagerouter.py [hunk 3]
```diff
@@ -1044,7 +1041,7 @@ def remove_storagedriver(storagedriver_guid, offline_storage_router_guids=None):
             raise RuntimeError('Could not found any responsive node in the cluster')
 
         storage_driver.invalidate_dynamics('vdisks_guids')
-        if len(storage_driver.vdisks_guids) > 0:
+        if len(storage_driver.vdisks_guids) > 0 and storage_router_online is True:
```
Comment (khenderick) [inline ovs/lib/storagerouter.py]: It should raise even if the StorageRouter is offline

File: scripts/system/ovs [hunk 4]
```diff
@@ -41,15 +41,15 @@ function show_help {
...
+    echo "    - ovs remove node <ip>            Remove node from cluster"
     echo ""
     echo "  * Monitor options:"
     echo "    - ovs monitor services            Watch Open vStorage services"
     echo "    - ovs monitor heartbeat           Send an internal heartbeat"
     echo ""
     echo "  * Services options:"
-    echo "    - ovs stop framework [ip|all]     Stop Open vStorage Framework services on this node, on all nodes, or on the given ip"
-    echo "    - ovs start framework [ip|all]    Start Open vStorage Framework services on this node, on all nodes, or on the given ip"
+    echo "    - ovs stop framework <ip|all>     Stop Open vStorage Framework services on this node, on all nodes, or on the given ip"
```
Comment (khenderick) [inline scripts/system/ovs]: Should be `[` and `]` as this implies an optional parameter.

**Review by khenderick — CHANGES_REQUESTED** [review]

**Review by khenderick — APPROVED** [review]

**Comment by kvanhijf** [comment]
https://github.com/openvstorage/framework/pull/1078 --> openvstorage-2.7.4-rev.4193.aa0ed1c
https://github.com/openvstorage/alba-asdmanager/pull/131 --> openvstorage-sdm-1.6.4-rev.394.22c93f2
https://github.com/openvstorage/framework-alba-plugin/pull/250 --> openvstorage-backend-1.7.4-rev.784.647dd17

**Pull request merged**